drivers/platform/nintendo3ds/ctr_gpio.c

```diff
@@ -0,0 +1,225 @@
+#include <linux/mod_devicetable.h>
+#include <linux/irqchip/chained_irq.h>
+
+struct ctr_gpio {
+	struct device *dev;
+
+	unsigned ngpios;
+	void __iomem *dat;
+	void __iomem *dir;
+	void __iomem *irqedge;
+	void __iomem *irqenable;
+
+	struct gpio_chip gpioc;
+	struct irq_chip irqc;
+};
+
+static void ctr_gpio_irqhandler(struct irq_desc *desc)
+{
+	int i, irq;
+	unsigned long flags, pending;
+	struct ctr_gpio *gpio =
+		gpiochip_get_data(irq_desc_get_handler_data(desc));
+	struct irq_chip *chip = irq_desc_get_chip(desc);
+
+	pending = 0;
+
+	spin_lock_irqsave(&gpio->gpioc.bgpio_lock, flags);
+	for (i = 0; i < (gpio->ngpios / 8); i++) {
+		u8 data, edge, enabled;
+		data = ioread8(gpio->dat + i);
+		edge = ioread8(gpio->irqedge + i);
+		enabled = ioread8(gpio->irqenable + i);
+		pending |= (~(data ^ edge) & enabled) << (8 * i);
+	}
+	spin_unlock_irqrestore(&gpio->gpioc.bgpio_lock, flags);
+
+	chained_irq_enter(chip, desc);
+	for_each_set_bit(irq, &pending, gpio->ngpios) {
+		generic_handle_irq(
+			irq_find_mapping(gpio->gpioc.irq.domain, irq));
+	}
+	chained_irq_exit(chip, desc);
+}
+
+static void ctr_gpio_irq_toggle(struct ctr_gpio *gpio,
+								unsigned irq, unsigned enable)
+{
+	u8 mask;
+	unsigned offset;
+	unsigned long flags;
+
+	offset = irq / 8;
+
+	spin_lock_irqsave(&gpio->gpioc.bgpio_lock, flags);
+	mask = ioread8(gpio->irqenable + offset);
+	mask = enable ? (mask | BIT(irq % 8)) : (mask & ~BIT(irq % 8));
+	iowrite8(mask, gpio->irqenable + offset);
+	spin_unlock_irqrestore(&gpio->gpioc.bgpio_lock, flags);
+}
+
+static void ctr_gpio_irq_mask(struct irq_data *data)
+{
+	struct ctr_gpio *gpio =
+		gpiochip_get_data(irq_data_get_irq_chip_data(data));
+	ctr_gpio_irq_toggle(gpio, data->hwirq, 0);
+}
+
+static void ctr_gpio_irq_unmask(struct irq_data *data)
+{
+	struct ctr_gpio *gpio =
+		gpiochip_get_data(irq_data_get_irq_chip_data(data));
+	ctr_gpio_irq_toggle(gpio, data->hwirq, 1);
+}
+
+static int ctr_gpio_irq_set_type(struct irq_data *data, unsigned int type)
+{
+	u8 mask;
+	unsigned long flags;
+	unsigned irq, offset;
+	struct ctr_gpio *gpio =
+		gpiochip_get_data(irq_data_get_irq_chip_data(data));
+
+	if (type != IRQ_TYPE_EDGE_RISING && type != IRQ_TYPE_EDGE_FALLING)
+		return -EINVAL;
+
+	irq = data->hwirq;
+	offset = irq / 8;
+
+	spin_lock_irqsave(&gpio->gpioc.bgpio_lock, flags);
+	mask = ioread8(gpio->irqedge + offset);
+
+	if (type == IRQ_TYPE_EDGE_RISING) {
+		mask |= BIT(irq % 8);
+	} else {
+		mask &= ~BIT(irq % 8);
+	}
+
+	iowrite8(mask, gpio->irqedge + offset);
+	spin_unlock_irqrestore(&gpio->gpioc.bgpio_lock, flags);
+	return 0;
+}
+
+static int ctr_gpiointc_probe(struct platform_device *pdev)
+{
+	u32 ngpios;
+	struct device *dev;
+	int err, i, irq_count;
+	struct ctr_gpio *gpio;
+	unsigned bgpio_flags, nregs;
+
+	dev = &pdev->dev;
+
+	gpio = devm_kzalloc(dev, sizeof(*gpio), GFP_KERNEL);
+	if (!gpio)
+		return -ENOMEM;
+
+	gpio->dev = dev;
+	gpio->dat = devm_platform_ioremap_resource(pdev, 0);
+	if (IS_ERR(gpio->dat))
```

Comment:
do we need to free `gpio` on error?

Reply:
No, everything is allocated via the managed resource API, whenever the driver exists (be it error or success) it'll release all the resources allocated with the devm_* functions.